modules/siddhi-core/src/main/java/org/wso2/siddhi/core/ExecutionPlanRuntime.java

```diff
@@ -257,24 +268,43 @@ public synchronized SiddhiDebugger debug() {
         return siddhiDebugger;
     }
 
-    public String persist() {
-        return executionPlanContext.getPersistenceService().persist();
+    public Future persist() {
+        try {
+            // first, pause all the event sources
+            eventSourceMap.values().forEach(list -> list.forEach(InputTransport::pause));
+            // take snapshots of execution units
+            byte[] snapshots = executionPlanContext.getSnapshotService().snapshot();
+            // start the snapshot persisting task asynchronously
+            return executionPlanContext.getExecutorService().submit(new AsyncSnapshotPersistor(snapshots,
+                    executionPlanContext.getSiddhiContext().getPersistenceStore(), executionPlanContext.getName()));
+        } finally {
+            // at the end, resume the event sources
+            eventSourceMap.values().forEach(list -> list.forEach(InputTransport::resume));
+        }
     }
 
     public void restoreRevision(String revision) {
-        executionPlanContext.getPersistenceService().restoreRevision(revision);
+        try {
+            // first, pause all the event sources
+            eventSourceMap.values().forEach(list -> list.forEach(InputTransport::pause));
+            // start the restoring process
+            executionPlanContext.getPersistenceService().restoreRevision(revision);
+        } finally {
+            // at the end, resume the event sources
+            eventSourceMap.values().forEach(list -> list.forEach(InputTransport::resume));
+        }
     }
 
     public void restoreLastRevision() {
-        executionPlanContext.getPersistenceService().restoreLastRevision();
-    }
-
-    public byte[] snapshot() {
```

Comment:
Some of the open source clients use snapshot() and restore(), hence please make that available too.